crates/red_knot/src/workspace/files.rs

```diff
@@ -0,0 +1,242 @@
+use std::iter::FusedIterator;
+use std::ops::Deref;
+use std::sync::Arc;
+
+use rustc_hash::FxHashSet;
+
+use crate::db::Db;
+use crate::workspace::Package;
+use ruff_db::files::File;
+
+/// The indexed files of a package.
+///
+/// The indexing happens lazily, but the files are then cached for subsequent reads.
+///
+/// ## Implementation
+/// The implementation uses internal mutability to transition between the lazy and indexed state
+/// without triggering a new salsa revision. The internal mutability **must not** be used for any
+/// other state transition that requires invalidating dependent salsa queries (e.g. adding or removing
+/// files from the index **must** go through the salsa setter to let salsa know about the input change).
```

Comment:
Can we make this comment clearer and more concrete? I think I understand it now, after poring over this PR for a while, but without having taken that time I think I would be scared by this comment but confused about how to actually follow it correctly.

I think that the practical upshot is that changes to `PackageFiles` (and the `IndexedFiles` that it references via its `State`) must only be made via `IndexedFilesMut`, which takes care to go through `Package::set_file_set` (the Salsa setter) to update to a new `PackageFiles` with updated revision, so that Salsa knows about the change.

```suggestion
/// without triggering a new salsa revision. This is safe because the initial indexing happens on first access,
/// so no query can be depending on the contents of the indexed files before that. All subsequent mutations to
/// the indexed files must go through `IndexedFilesMut`, which uses the Salsa setter `package.set_file_set` to
/// ensure that Salsa always knows when the set of indexed files have changed.
```